src/dune/build_system.ml

```diff
@@ -1558,8 +1639,18 @@ end = struct
                 in
                 Log.info [ Pp.textf "promotion failed for %s: %s" targets msg ]
               in
+              let repository =
+                let dir = find_rule_source_dir rule in
+                match File_tree.Dir.vcs dir with
+                | Some vcs -> (
+                  let f found = Path.equal found.Vcs.root vcs.Vcs.root in
+                  match get_vcs () |> List.findi ~f with
+                  | Some (_, i) -> Some i
+                  | None -> None )
+                | None -> None
```

Comment:
This could be cleaned with binds:

```
let open Option.O in
let* vcs = File_tree.Dir.vcs dir with in
let f found = Path.equal found.Vcs.root vcs.Vcs.root in
let+ (_, i) = get_vcs () |> List.findi ~f in
i
```